packages/aws-cdk/lib/api/util/cloudformation.ts

```diff
@@ -367,19 +367,19 @@ function templatesFromAssetManifestArtifact(artifact: cxapi.AssetManifestArtifac
 async function uploadBodyParameterAndCreateChangeSet(options: PrepareChangeSetOptions): Promise<DescribeChangeSetOutput | undefined> {
   try {
     await uploadStackTemplateAssets(options.stack, options.deployments);
-    const preparedSdk = (await options.deployments.prepareSdkWithDeployRole(options.stack));
+    const env = await options.deployments.envs.accessStackForStackOperations(options.stack, Mode.ForWriting);
 
     const bodyParameter = await makeBodyParameter(
       options.stack,
-      preparedSdk.resolvedEnvironment,
+      env.resolvedEnvironment,
       new AssetManifestBuilder(),
-      preparedSdk.envResources,
-      preparedSdk.stackSdk,
+      env.resources,
+      env.sdk,
     );
-    const cfn = preparedSdk.stackSdk.cloudFormation();
+    const cfn = env.sdk.cloudFormation();
     const exists = (await CloudFormationStack.lookup(cfn, options.stack.stackName, false)).exists;
 
-    const executionRoleArn = preparedSdk.cloudFormationRoleArn;
+    const executionRoleArn = await env.replacePlaceholders(options.stack.cloudFormationExecutionRoleArn);
```

Comment:
It used to be: you must pass in the role ARN immediately when specializing for an environment, and then it would replace placeholders inside that ARN if present. This also meant you had to pass an `undefined` in case you didn't actually need placeholders replaced inside that ARN.

The current solution of doing conversion if and when you need it seems better.